x-pack/apm-server/sampling/eventstorage/storage_manager.go

```diff
@@ -26,315 +27,335 @@ const (
+	// This means storage requirement is 2 * TTL, and it needs to read 2 keys per trace ID read.
+	// If partitionPerTTL=2, storage requirement is 1.5 * TTL at the expense of 3 reads per trace ID read.
+	partitionsPerTTL = 1
+
+	// reservedKeyPrefix is the prefix of internal keys used by StorageManager
+	reservedKeyPrefix byte = '~'
+
+	// partitionerMetaKey is the key used to store partitioner metadata, e.g. last partition ID, in decision DB.
+	partitionerMetaKey = string(reservedKeyPrefix) + "partitioner"
+
+	// diskUsageFetchInterval is how often disk usage is fetched which is equivalent to how long disk usage is cached.
+	diskUsageFetchInterval = 1 * time.Second
 )
 
-// StorageManager encapsulates badger.DB.
-// It is to provide file system access, simplify synchronization and enable underlying db swaps.
-// It assumes exclusive access to badger DB at storageDir.
+type StorageManagerOptions func(*StorageManager)
+
+func WithCodec(codec Codec) StorageManagerOptions {
+	return func(sm *StorageManager) {
+		sm.codec = codec
+	}
+}
+
+func WithMeterProvider(mp metric.MeterProvider) StorageManagerOptions {
+	return func(sm *StorageManager) {
+		sm.meterProvider = mp
+	}
+}
+
+// StorageManager encapsulates pebble.DB.
+// It assumes exclusive access to pebble DB at storageDir.
 type StorageManager struct {
 	storageDir string
 	logger     *logp.Logger
 
-	db      *badger.DB
-	storage *Storage
-	rw      *ShardedReadWriter
+	eventDB         *pebble.DB
+	decisionDB      *pebble.DB
+	eventStorage    *Storage
+	decisionStorage *Storage
+
+	partitioner *Partitioner
+
+	storageLimit atomic.Uint64
+
+	codec Codec
 
-	// mu guards db, storage, and rw swaps.
-	mu sync.RWMutex
 	// subscriberPosMu protects the subscriber file from concurrent RW.
 	subscriberPosMu sync.Mutex
 
+	// cachedDiskUsage is a cached result of DiskUsage
+	cachedDiskUsage atomic.Uint64
+
 	// runCh acts as a mutex to ensure only 1 Run is actively running per StorageManager.
 	// as it is possible that 2 separate Run are created by 2 TBS processors during a hot reload.
 	runCh chan struct{}
+
+	// meterProvider is the OTel meter provider
+	meterProvider metric.MeterProvider
+
+	metricRegistration metric.Registration
 }
 
-// NewStorageManager returns a new StorageManager with badger DB at storageDir.
-func NewStorageManager(storageDir string) (*StorageManager, error) {
+// NewStorageManager returns a new StorageManager with pebble DB at storageDir.
+func NewStorageManager(storageDir string, opts ...StorageManagerOptions) (*StorageManager, error) {
 	sm := &StorageManager{
 		storageDir: storageDir,
 		runCh:      make(chan struct{}, 1),
 		logger:     logp.NewLogger(logs.Sampling),
+		codec:      ProtobufCodec{},
 	}
-	err := sm.reset()
-	if err != nil {
-		return nil, err
+	for _, opt := range opts {
+		opt(sm)
+	}
+
+	if sm.meterProvider != nil {
+		meter := sm.meterProvider.Meter("github.com/elastic/apm-server/sampling/eventstorage")
```

Comment:
Shouldn't it be `meter := sm.meterProvider.Meter("github.com/elastic/apm-server/x-pack/apm-server/sampling/eventstorage")`?

Reply:
Nice catch. Shouldn't code at night.